event-dispatcher/src/main/java/io/sk8s/event/dispatcher/EventDispatchingHandler.java

```diff
@@ -16,55 +16,264 @@
+					functions.values().stream().forEach(
+							f -> {
+								int desired = computeDesiredNbReplicas(lags, f);
+								String name = f.getMetadata().getName();
+								Integer currentDesired = actualNbReplicas.computeIfAbsent(name, k -> desired);
+								System.out.println("Want " + desired + " for " + name);
+								if (desired < currentDesired) {
+									if (edgeInfos.computeIfAbsent(name,
+											n -> System.currentTimeMillis()) < System.currentTimeMillis()
+													- DOWN_EDGE_LINGER_PERIOD) {
+										deployer.deploy(f, desired);
+										actualNbReplicas.put(name, desired);
+										edgeInfos.remove(name);
+									}
+									else {
+										System.out.printf("Waiting another %dms before scaling down to %d for %s%n",
+												DOWN_EDGE_LINGER_PERIOD
+														- (System.currentTimeMillis() - edgeInfos.get(name)),
+												desired, name);
+									}
+								}
+								else if (desired > currentDesired) {
+									deployer.deploy(f, desired);
+									actualNbReplicas.put(name, desired);
+									edgeInfos.remove(name);
+								}
+								else {
+									edgeInfos.remove(name);
+								}
+							});
+
+					try {
+						EventDispatchingHandler.this.wait(scalingFrequencyMs);
+					}
+					catch (InterruptedException e) {
+						Thread.currentThread().interrupt();
+					}
+				}
+			}
+		}
+
+		/**
+		 * Compute the desired nb of replicas for a function. Each function defines 4 values:
+		 * <ul>
+		 * <li>minReplicas (>= 0, default 0)</li>
+		 * <li>maxReplicas (minReplicas <= maxReplicas <= nbPartitions, default nbPartitions)</li>
+		 * <li>l, the amount of lag required to trigger the first pod to appear, default 1</li>
+		 * <li>L, the amount of lag required to trigger all (maxReplicas) pods to appear. Default
+		 * 10.</li>
+		 * </ul>
+		 * This method linearly interpolates based on witnessed lag and clamps the result between
+		 * min/maxReplicas.
+		 */
+		private int computeDesiredNbReplicas(Map<String, Long> lags, XFunction f) {
+			String fnName = f.getMetadata().getName();
+			long lag = lags.get(fnName);
+
+			// TODO: those 3 numbers part of Function spec?
+			int bigL = 10;
+			int littleL = 1;
+			int min = 0;
+
+			String input = f.getSpec().getInput();
+			double max = nbPartitions(input);
+			double slope = (max - 1) / (bigL - littleL);
+			int computedReplicas;
+			if (slope > 0d) {
+				// max>1
+				computedReplicas = (int) (1 + (lag - littleL) * slope);
+			}
+			else {
+				computedReplicas = lag >= littleL ? 1 : 0;
+			}
+			return clamp(computedReplicas, min, (int) max);
+		}
+
+		private int nbPartitions(String input) {
+			waitForTopicEntry(input);
+			TopicSpec spec = topics.get(input).getSpec();
+			if (spec == null || spec.getPartitions() == null) {
+				return 1;
+			}
+			else {
+				return spec.getPartitions();
+			}
+		}
+
+		private void waitForTopicEntry(String input) {
+			try {
+				topicsReady.computeIfAbsent(input, i -> new CountDownLatch(1)).await();
```

Comment:
seems like we might want a bounded wait timeout here (also I'd drop "Entry" from the method name... possibly "waitForTopicIfNecessary"?)

Reply:
What if we timeout? A NPE or similar will happen elsewhere. Absence of entry should never happen (famous last words)

Reply:
as long as you're confident this can't lead to deadlock